JCL-406: JSON metadata deserialization

NSeydoux opened the pull request:
Forked from https://github.com/inrupt/solid-client-java/pull/550, using a default naming strategy that initially failed locally but looks like it is working now, which looks like misusage of the IDE on my end.

NSeydoux committed:
Make mapping class public

JSONB complains otherwise.

NSeydoux committed:
Add bigint support for status credential parsing

In some cases, JSONB instantiates integers as BigInt instances, which wasn't accounted for in the cast to integer.

NSeydoux committed:
Remove inner verification status class

Instead of having a mutable verification status class (which was intended to be private, but JSONB complained about that), the public AccessCredentialVerification class has now public setters and a default constructor to be usable directly by the JSON binding service. The setters will make the attributes immutable, so that they can only be set once.

NSeydoux committed:
Re-enable failing tests

NSeydoux committed:
Implement appropriate JSONB naming strategy

Jsonb current naming strategy fails to correctly parse UMA metadata, or any snake case JSON key that should map to camelcased class attributes starting with a lowercase. This new naming strategy resolves the issue.

NSeydoux committed:
Revert "Implement appropriate JSONB naming strategy"

This reverts commit 0000bd3af7b931b8e2be906215661cbd837bcd56.

NSeydoux committed:
Make snake case json-b default naming strategy

Aligns with Jackson default strategy.

NSeydoux committed:
Enable JSON-B tests on core module

NSeydoux committed:
Enable JSON-B tests on openid module

NSeydoux committed:
Re-enabled JSON-B tests in UMA module

NSeydoux commented on jsonb/src/main/java/com/inrupt/client/jsonb/JsonbService.java:
```diff
@@ -46,7 +46,7 @@ public class JsonbService implements JsonService {
     public JsonbService() {
         final JsonbConfig config = new JsonbConfig();
-        config.withPropertyNamingStrategy(PropertyNamingStrategy.LOWER_CASE_WITH_DASHES);
+        config.withPropertyNamingStrategy(PropertyNamingStrategy.LOWER_CASE_WITH_UNDERSCORES);
```
That is correct. The default for JSONB is `IDENTITY`, which doesn't change property names between Java and JSON, so the initial state of this module was already a deviation from the default. As you point out, this change aligns this with the Jackson module behaviour.

NSeydoux commented on access-grant/src/main/java/com/inrupt/client/accessgrant/AccessCredentialVerification.java:
```diff
@@ -28,9 +28,18 @@
+    /**
+     * Create an empty verification response.
+     */
+    public AccessCredentialVerification() {
```
This class is indeed only used for deserializing responses from the `/verify` endpoint. It was initially immutable, with a parametrized constructor and `final` members. However, this relied on `VerificationResponseData`, a class private to `AccessGrantClient`. Due to JSON-B introspection restrictions, `VerificationResponseData` should be public, which is undesirable. The current implementation of `AccessCredentialVerification` works around this issue as you describe, by exposing a default constructor (as required by JSON-B, since we aren't using annotations), and still making the members immutable once they have been initialized.

I can't tell if it is a hard requirement that `AccessCredentialVerification` instances are immutable. It makes sense semantically, but I don't know how crucial it is vs implementing less circonvoluted patterns.

langsamu replied:
I see.

The class proposed here would not be truely immutable.
The collections its members return _are_ immutable, which is good.
However the collections _themselves_ can be overwritten with the new setters, which is not good.

How about the following pattern:
There is a class `VerificationResponseData` (as before). It's a pure DTO, a POJO with public, writable fields that is used for JSON deserialization only. That class itself must be public for JSONB, which is undesirable, as you say. We document it for internal user only.

The `AccessCredentialVerification` class, on the other hand, is a wrapper around the DTO (as before), that exposes only getters that make the fields immutable.
That way you return a truely immutable, clean API class.

The only disadvantage is exposing the DTO, but I think it's worth it. Maybe even suppress the warning for public fields and mutable collections.

Something like
```java
/**
 * for internal use only
 */
public class VerificationResponseData {
  /**
   * for internal use only
   */
  public List<String> checks;

  // etc
}

/**
 * properly documented
 */
public class AccessCredentialVerification {
  private final VerificationResponseData  original;

  /* not even public */ AccessCredentialVerification(final VerificationResponseData original) {
    this.original = original;
  }

  /**
   * properly documented
   */
  public List<String> getChecks() {
    return Collections.unmodifiableList(original.checks);
  }
}
```

NSeydoux replied:
Based on the conversation in https://github.com/inrupt/solid-client-java/pull/542, my previous comment was mislead: it is not a hard requirement that the verification data class is immutable. I think exposing internals in the public API requires more care from us when refactoring the code, so let's just make the class mutable.
However, since the current gimmick of having immutable collections in mutable attributes doesn't really follow any well-established pattern, let's just make everything mutable for consistency, so that it isn't confusing.

NSeydoux committed:
Simplify AccessCredentialVerification

AccessCredentialVerification was using a non-standard pattern of immutability of the collections accessed by attributes, while the attributes themselves were not immutable. This makes everything mutable to have a more standard data class.

NSeydoux committed:
fixup! Simplify AccessCredentialVerification

langsamu reviewed (approved):
[Formatting complaint](https://github.com/inrupt/solid-client-java/pull/559/files#diff-4751c4556f19f86b0a6b622a3fe9e4eab7077b647b7182a5623c8832a69d73efR91)
